src/canisters/frontend/ic-certified-assets/src/state_machine.rs

```diff
@@ -678,59 +823,113 @@ fn create_token(
     }
 }
 
-#[allow(clippy::too_many_arguments)]
-fn build_ok(
-    asset: &Asset,
-    enc_name: &str,
-    enc: &AssetEncoding,
-    key: &str,
-    chunk_index: usize,
-    certificate_header: Option<HeaderField>,
-    callback: Func,
-    etags: Vec<Hash>,
-) -> HttpResponse {
-    let mut headers = vec![("Content-Type".to_string(), asset.content_type.to_string())];
-    if enc_name != "identity" {
-        headers.push(("Content-Encoding".to_string(), enc_name.to_string()));
-    }
-    if let Some(head) = certificate_header {
-        headers.push(head);
+impl RedirectUrl {
+    fn match_and_replace_segment(
+        match_segment: &Option<String>,
+        replace_with: &Option<String>,
+        in_text: &str,
+    ) -> Option<String> {
+        if let Some(replace) = replace_with {
+            if match_segment.is_none() {
+                return Some(replace.to_owned());
+            }
+            if let Ok(regex) = Regex::new(match_segment.as_ref().unwrap()) {
+                return regex
+                    .find(in_text)
+                    .map(|_| regex.replace(in_text, replace).to_string());
+            }
+        }
+        None
     }
-    if let Some(max_age) = asset.max_age {
-        headers.push(("Cache-Control".to_string(), format!("max-age={}", max_age)));
+}
+
+impl HttpRequest {
+    fn get_header_value(&self, header_key: &str) -> Option<&String> {
+        self.headers
+            .iter()
+            .find_map(|(k, v)| k.eq_ignore_ascii_case(header_key).then(|| v))
     }
-    if let Some(arg_headers) = asset.headers.as_ref() {
-        for (k, v) in arg_headers {
-            headers.push((k.to_owned(), v.to_owned()));
+}
+
+impl HttpResponse {
```

Comment:
I did a small refactor, and have put all the different HTTP responses under one umbrella